[NEW] Sections layout and featured apps for marketplace

rique223 opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

<!-- Your Pull Request name should start with one of the following tags
  [NEW] For new features
  [IMPROVE] For an improvement (performance or little improvements) in existing features
  [FIX] For bug fixes that affect the end-user
  [BREAK] For pull requests including breaking changes
  Chore: For small tasks
  Doc: For documentation
-->

<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)
<!-- CHANGELOG -->
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description will appear in the release notes if we accept the contribution.
-->
### [MKP??+ - Featured Apps](https://app.clickup.com/t/2t1w9x3)
Implemented a new layout for marketplace's apps list page. Now the page has an All apps section and a "featured apps" section in which pre-selected and pre-categorized apps are dynamically shown on the top area of the page to make them stand out.
Demo gif:
![featured_apps](https://user-images.githubusercontent.com/43561537/184385727-4f44c1d5-d7f7-43bf-a93d-2bc582bd3159.gif)

ClickUp Task link: https://app.clickup.com/t/2t1w9x3
<!-- END CHANGELOG -->

## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->

## Steps to test or reproduce
<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

rique223 committed:
feat: :sparkles: Surface featured apps endpoint to RC

Created the /featured endpoints on the rest.js file. Also created the necessary typings to use together with it.

rique223 committed:
Merge remote-tracking branch 'origin/develop' into feat/surface-featured-apps-endpoint

rique223 committed:
Review

rique223 committed:
fix: :bug: Fix priority for featured apps endpoint and typo

Fixed a problem in which the featured endpoint would be out prioritized by the rest client because it was bellow a wildcard endpoint. Also solved a typo on the route definition for the featured endpoint.

rique223 committed:
Remove unecessary console.log

rique223 committed:
WIP: Generalize AppsList

rique223 committed:
chore: :recycle: Refactor AppList to be generic for App arrays

Refactored the AppsList and surrounding components so that they can be reused for more than one list of apps and start the creation of the new sections.

rique223 committed:
feat: :sparkles: Consume featured endpoint and show new sections

Consumed the response from the /featured endpoint, created a normalization function for the apps list of the sections and created the necessary types to make it all work.

rique223 committed:
Merge remote-tracking branch 'origin/develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
feat: :sparkles: Implement search and loading behavior for new featured apps sections

Implemented the on search and on load behavior for the new featured apps sections and solved some missing key problems.

rique223 committed:
Merge branch 'feat/featured-apps-marketplace' of github.com:RocketChat/Rocket.Chat into feat/featured-apps-marketplace

rique223 committed:
Merge remote-tracking branch 'origin/develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
feat: :sparkles: Hide featured app sections on filter

Implemented the logic for the featured apps sections to hide when the filters are used

rique223 committed:
Remove unnecessary console.log

rique223 committed:
style: :lipstick: Restyle marketplace apps page tabs

Changed the styles of the marketplace apps page tabs to make it closer to the figma design

ggazzo commented on apps/meteor/client/views/admin/apps/AppsList.tsx:
```diff
@@ -1,202 +1,26 @@
+} & ComponentProps<typeof Box>;
+
+const AppsList = ({ apps, title, isMarketplace, ...props }: AppsListProps): ReactElement => (
+	<Fragment key={props.key}>
```
I'm quite sure prop.key will throw you a console.warning

rique223 replied:
Solved

ggazzo commented on apps/meteor/client/views/admin/apps/AppsPage.tsx:
```diff
@@ -58,16 +58,29 @@ const AppsPage = ({ isMarketplace }: AppsPageProps): ReactElement => {
 			</Page.Header>
-			<Tabs bg={colors.n100}>
-				<Tabs.Item onClick={(): void => marketplaceRoute.push({ context: '' })} selected={isMarketplace}>
+			<Tabs bg={colors.n100} borderBlockEnd='4px solid #CBCED1' marginInline='x24'>
```
dont just, dont  pass styles to components, if you want, create this divider inside the fuselage lib and than use as some kind of variation

rique223 replied:
Will open a task to solve this since waiting for fuselage would unnecessarily stall this PR.

ggazzo commented on apps/meteor/client/views/admin/apps/AppsPageContent.tsx:
```diff
@@ -0,0 +1,213 @@
+									{t('No_app_matches_for')} <strong>"{text}"</strong>
+								</span>
+							</StatesSubtitle>
+						) : (
+							''
+						)}
```
🤔 is that really necessary?

rique223 replied:
Solved

ggazzo reviewed (dismissed):
its not a good practice put a bunch of <Box ....props> around you application code.
its recommended to create a 'custom component' like "FeaturedAppSectionItem" and them there you are allowed to use some styling

rique223 committed:
refactor: :recycle: Change featured endpoint and solve reviews

The url of the featured endpoint was causing a problem on the back-end, changed the url to v1/featured-apps to solve this problem. Also solved Gazzos reviews.

rique223 committed:
Merge remote-tracking branch 'origin/develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

d-gubert commented on apps/meteor/app/apps/client/orchestrator.ts:
```diff
@@ -107,15 +107,10 @@ class AppClientOrchestrator {
-				isEnterpriseOnly,
-				modifiedAt,
-				bundledIn,
+				...rest,
```
IIRC `latest` has some properties with the same name as the ones in the top-level `app` object, so spreading the `rest` here will likely override those if done this way.

I also like extracting exactly the info we need from the original `app` object so that we don't "overshare" unnecessary data, but that is not objectively better (unlike the above), just my two cents :P

rique223 replied:
Got it, will solve this shortly. I think we should take the refactoring of this app object to the engineering day. Right now it is a simple spread but it can become a big problem in the future.

rique223 replied:
Solved

d-gubert commented on apps/meteor/app/apps/server/communication/rest.js:
```diff
@@ -362,7 +362,7 @@ export class AppsRestApi {
 					let result;
 					try {
-						result = HTTP.get(`${baseUrl}/v1/apps/featured`, {
+						result = HTTP.get(`${baseUrl}/v1/featured-apps`, {
```
This is the URL in the Marketplace API, not the Rocket.Chat server. Please double check if the Marketplace API has introduced this new endpoint, as my test with cURL (`curl https://marketplace.rocket.chat/v1/featured-apps`) hints it doesn't exist

rique223 replied:
This endpoint is still in the marketplace staging, it will be released shortly.

rique223 replied:
This is the staging URL if you want to test it: https://marketplace.staging.cloud.rocket.chat/v1/featured-apps

juliajforesti commented on apps/meteor/client/views/admin/apps/BundleChips.tsx:
```diff
@@ -22,7 +22,7 @@ const BundleChips = ({ bundledIn, isIconOnly }: BundleChipsProps): ReactElement
 		<>
 			{bundledIn.map((bundle) => (
-				<>
+				<Fragment key={bundle.bundleId}>
```
Comment for line 41:
try using our typography tokens from [fuselage](http://localhost:6006/?path=/docs/layout-box-props--font-sizes) to keep the standards.
In this case would be **c2**

rique223 replied:
Solved

rique223 committed:
Reviews

rique223 committed:
Merge branch 'feat/featured-apps-marketplace' of github.com:RocketChat/Rocket.Chat into feat/featured-apps-marketplace

rique223 commented on apps/meteor/client/views/admin/apps/AppsPage.tsx:
```diff
@@ -58,16 +58,29 @@ const AppsPage = ({ isMarketplace }: AppsPageProps): ReactElement => {
 					)}
 				</ButtonGroup>
 			</Page.Header>
-			<Tabs bg={colors.n100}>
```
Solved

rique223 commented on apps/meteor/client/views/admin/apps/AppsPageContent.tsx:
```diff
@@ -0,0 +1,209 @@
+					count={appsResult.value.total}
+					onSetItemsPerPage={onSetItemsPerPage}
+					onSetCurrent={onSetCurrent}
+					borderBlockStart={`2px solid ${colors.n300}`}
```
Solved

rique223 committed:
Merge remote-tracking branch 'origin/develop' into feat/featured-apps-marketplace

rique223 committed:
Reviews

rique223 commented on apps/meteor/client/views/admin/apps/AppsPageContent.tsx:
```diff
@@ -0,0 +1,209 @@
+			)}
+
+			{noInstalledAppMatches && (
+				<Box mbs='x20'>
```
Solved

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

ggazzo commented on apps/meteor/client/views/admin/apps/AppsList.tsx:
```diff
@@ -1,202 +1,26 @@
+
+const AppsList = ({ apps, title, isMarketplace, ...props }: AppsListProps): ReactElement => (
+	<>
+		<Box is='h3' fontSize='h3' fontWeight={700} lineHeight='x28' color='default' mbe='x20'>
```
instead of `fontSize='h3' fontWeight={700} lineHeight='x28'` I'm sure that exists some fontScale to be used

rique223 replied:
Solved

rique223 committed:
fix: :bug: Pagination bug and reviews

Solved a bug on the apps list pagination in which the pagination component was appearing in the middle of the all apps list instead of at the end.

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

matheuslc commented on packages/core-typings/src/AppOverview.ts:
```diff
@@ -0,0 +1,115 @@
+	addonTierId: string;
+};
+
+export type Latest = {
```
@rique223 as I understood, the "latest" version could be different from workspaces. It depends on which engine version they are using. So, what do you think about renaming this object to CurrentVersion? This way we do not assume we are always using the Latest version of an app.

rique223 replied:
Solved

rique223 committed:
refactor: :recycle: AppsList loading skeletons behavior

Changed the behavior of the skeletons of the loading state for the AppsList, now the skeletons only go away when its possible to render the whole screen.

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

casalsgh committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

graywolf336 commented on apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:
```diff
@@ -3082,6 +3082,8 @@
   "Marketplace_app_last_updated": "Last updated __lastUpdated__",
   "Marketplace_view_marketplace": "View Marketplace",
   "Marketplace_error": "Cannot connect to internet or your workspace may be an offline install.",
+  "marketplace_featured_section_featured": "Featured apps",
+  "marketplace_featured_section_social": "New arrivals",
```
```suggestion
  "marketplace_featured_section_community_featured": "Featured Community Apps",
  "marketplace_featured_section_community_supported": "Community Supported Apps",
  "marketplace_featured_section_enterprise": "Featured Enterprise Apps",
  "marketplace_featured_section_featured": "Featured Apps",
  "marketplace_featured_section_most_popular": "Most Popular Apps",
  "marketplace_featured_section_new_arrivals": "New Arrivals",
  "marketplace_featured_section_popular_this_month": "Apps Popular this Month",
  "marketplace_featured_section_recommended": "Recommended Apps",
  "marketplace_featured_section_social": "Social Apps",
  "marketplace_featured_section_trending": "Trending Apps",
  "marketplace_featured_section_omnichannel": "Omnichannel Apps",
  "marketplace_featured_section_video_conferencing": "Video Conferencing Apps",
```

graywolf336 reviewed (changes requested):
Let's add these sections :)

rique223 committed:
Update apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json

Co-authored-by: Bradley Hilton <graywolf336@craftyn.com>

graywolf336 reviewed (dismissed):
Looks good to me and I've seen it running, from the Marketplace perspective it looks good and the APIs area in production to be used now.

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
refactor: :recycle: Component function definitions

Refactored some components definitions to better follow RCs standard.

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

casalsgh committed:
Merge branch 'develop' into feat/featured-apps-marketplace

juliajforesti commented on apps/meteor/client/views/admin/apps/components/RadioDropDown/RadioDropDown.tsx:
```diff
@@ -9,7 +9,7 @@ import DropDownListWrapper from '../DropDownListWrapper';
 import RadioDropDownAnchor from './RadioDownAnchor';
 
-const RadioDropDown: FC<RadioDropDownProps & Partial<ComponentProps<typeof Select>>> = ({ group, onSelected, ...props }) => {
+const RadioDropDown = ({ group, onSelected, ...props }: RadioDropDownProps & Partial<ComponentProps<typeof Select>>): ReactElement => {
```
I would suggest not to type with all Select props, but to define specifically the ones that are being used. Or is there a reason for forwarding all "rest" props? It increases the chance of unwanted/unforeseen props being passed.

rique223 replied:
The idea was for RadioDropDown to be a direct abstraction of Select that's why I used all props of Select so that both components can be used interchangeably.

juliajforesti commented on apps/meteor/client/views/admin/apps/FeaturedAppsSections.tsx:
```diff
@@ -0,0 +1,47 @@
+					<AppsList
+						key={section.slug}
+						apps={normalizeFeaturedApps(section.apps, appsResult.value.items)}
+						title={t(section.i18nLabel as TranslationKey)}
```
useTranslation has the method `has` that checks if this key exists.
```suggestion
title={t.has(section.i18nLabel) ? t(section.i18nLabel) : section.i18nLabel}
```

rique223 replied:
Solved

juliajforesti commented on apps/meteor/client/views/admin/apps/AppsList.tsx:
```diff
@@ -1,202 +1,26 @@
-		</>
-	);
-};
+} & ComponentProps<typeof Box>;
```
`ComponentProps<typeof Box>` is a whole lot of props... do you really want to add all that?

rique223 replied:
Solved

rique223 committed:
refactor: :recycle: Solve reviews

rique223 committed:
Merge branch 'feat/featured-apps-marketplace' of github.com:RocketChat/Rocket.Chat into feat/featured-apps-marketplace

rique223 committed:
Merge remote-tracking branch 'origin/develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

juliajforesti commented on apps/meteor/client/views/admin/apps/FeaturedAppsSections.tsx:
```diff
@@ -0,0 +1,47 @@
+						apps={normalizeFeaturedApps(section.apps, appsResult.value.items)}
+						title={t.has(section.i18nLabel) ? t(section.i18nLabel) : section.i18nLabel}
+						isMarketplace={isMarketplace}
+						mbe='x36'
```
is this prop necessary? 
would be better if you could put the margin on the parent component, not pass it as a prop

rique223 replied:
Solved

rique223 committed:
Review

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

d-gubert reviewed (dismissed):
All good on Apps side :)

geekgonecrazy committed:
Merge branch 'develop' into feat/featured-apps-marketplace

pierre-lehnen-rc commented on packages/rest-typings/src/apps/index.ts:
```diff
@@ -96,9 +96,9 @@ export type AppsEndpoints = {
 	};
 
-	'/apps/featured': {
+	'/apps/featured-apps': {
```
Isn't this a breaking change?

graywolf336 replied:
No sir. That route was added in https://github.com/RocketChat/Rocket.Chat/pull/26416 and was added specifically for this new feature. The Marketplace API doesn't have the old route after we tested in staging and realized it wouldn't work.

ggazzo committed:
review

rique223 committed:
fix: :bug: Fix pagination behavior, data badge component and dupped boolean

Fixed a bug in the Apps page pagination in which the pagination appeared in the installed page even though there were no installed apps. Also erased a duplicated boolean and solved a misalignment between the data badge component and the install/buy button.

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

rique223 committed:
Merge branch 'develop' into feat/featured-apps-marketplace

kodiakhq[bot] committed:
Merge branch 'develop' into feat/featured-apps-marketplace
